src/expertsystem/particle.py

```diff
@@ -208,6 +229,31 @@ def is_lepton(self) -> bool:
             or self.tau_lepton_number != 0
         )
 
+    def asdict(self) -> Any:
+        return attr.asdict(
+            self, recurse=True, value_serializer=value_serializer
+        )
+
+    @staticmethod
+    def fromdict(definition: dict) -> "Particle":
+        kwargs: Dict[str, Any] = dict()
+        for field in attr.fields(Particle):
+            item_definition = definition[field.name]
+            union_types = getattr(field.type, "__args__", None)
+            if union_types is not None:  # Union
+                attribute_type = union_types[0]
+            else:
+                attribute_type = field.type
+            if item_definition is None:
+                kwargs[field.name] = None
+            elif issubclass(attribute_type, Serializable):
+                kwargs[field.name] = attribute_type.fromdict(item_definition)
+            elif attr.has(attribute_type):
+                kwargs[field.name] = attribute_type(**item_definition)
+            else:
+                kwargs[field.name] = item_definition
```

Comment:
This seems a bit overly complicated. I would assume that `definition` is correct, and if not expect an error if its not. So I would have used the `definition` as the starting point for `kwargs`, and just replace the nested types (in other words, if `item_definition` is a `dict`).

Reply:
I gave iterating over the input definition a try, but the problem there is to determine what kind of constructor to call based on the value. I loop over the fields instead, because they contain more info.

The method is now created with a decorator 01e7a92 -- not really satisfied with it, but now the same method can be applied to other nested `attr.s` classes, such as `AmplitudeModel`.

Reply:
Note that the methods of the `Serializable` are not abstract anymore. I had to do this, because the decorator does not remove the abstract status (maybe this can be done dynamically though).

Did make me wonder whether a `Protocol` would be better here. That would also remove the need for the developer to specify that a class derives from `Serializable`.